examples/QIR/Simulation/StateSimulator/README.md

```diff
@@ -0,0 +1,336 @@
+    sort(postTargetIndices.begin(), postTargetIndices.end());
+
+    long dimU = 2;
+    Operator unitary = gate;
+    Operator project0 = (Operator(2,2) << 1,0,0,0).finished();
+    Operator project1 = (Operator(2,2) << 0,0,0,1).finished();
+    // Build up unitary from target to last qubit.
+    auto controlItFw = postTargetIndices.begin();
+    for (int i = targetIndex+1; i < this->numActiveQubits; i++) {
+        if (controlItFw != postTargetIndices.end() && i == *controlItFw) {
+            unitary = (kroneckerProduct(Operator::Identity(dimU, dimU), project0) // 1 ⊗ |0><0|
+                      +kroneckerProduct(unitary, project1)).eval();               // U ⊗ |1><1|
+            controlItFw++;
+        } else {
+            unitary = kroneckerProduct(unitary, Operator::Identity(2,2)).eval();
+        }
+        dimU *= 2;
+    }
+    // Build up the unitary from target to first qubit.
+    auto controlItBw = preTargetIndices.rbegin();
+    for (int i = targetIndex-1; i >= 0; i--) {
+        if (controlItBw != preTargetIndices.rend() && i == *controlItBw) {
+            unitary = (kroneckerProduct(project0, Operator::Identity(dimU, dimU)) // |0><0| ⊗ 1
+                      +kroneckerProduct(project1, unitary)).eval();               // |1><1| ⊗ U
+            controlItBw++;
+        } else {
+            unitary = kroneckerProduct(Operator::Identity(2,2), unitary).eval();
+        }
+        dimU *= 2;
+    }
+
+    // Apply gate with |Ψ'> = U|Ψ>.
+    this->stateVec = unitary*this->stateVec;
+}
+```
+
+We also need to define what happens to the state vector when we add or remove a qubit.
+In the case of adding a new qubit, the tensor product (or Kronecker product) is used to add the qubit to the state vector (last in the register, i.e `|Ψ'> = |Ψ> ⊗ |0>`).
+When removing a qubit, it is assumed to be in a product state with the rest of the register, and can thus be traced out from the state vector (i.e. `ρ' = |Ψ'><Ψ'| = tr_i[|Ψ><Ψ|]`).
+The `PartialTrace` method is defined in `StateSimulation.cpp` based on the definition `tr_B[U] = (Id ⊗ <0|) U (Id ⊗ |0>) + (Id ⊗ <1|) U (Id ⊗ |1>)`).
+The new state vector is then extracted from the density matrix using eigenvalue decomposition.
+Up to computational precision, there should only be a single eigenvalue whose corresponding eigenvector is the new state vector:
+
+```cpp
+void StateSimulator::UpdateState(short qubitIndex, bool remove)
+{
+    // When adding a qubit, the state vector can be updated with: |Ψ'> = |Ψ> ⊗ |0>.
+    // When removing a qubit, it is traced out from the state vector: ρ' = tr_i[|Ψ><Ψ|].
+    if (!remove) {
+        this->stateVec = kroneckerProduct(this->stateVec, Vector2cd(1,0)).eval();
+    } else {
+        Operator densityMatrix = this->stateVec * this->stateVec.adjoint();
+        densityMatrix = PartialTrace(densityMatrix, qubitIndex, this->numActiveQubits);
+
+        // Ensure state is pure tr(ρ^2)=1, meaning the removed qubit was in a product state.
```

Comment:
It'd be good to sync with @swernli on this to make sure that release semantics agree with those on other simulators.

Reply:
Reading through this, I think it works fine. It doesn't match the strict statement of released qubits needing to be measured or zero, but it does ensure that qubits are unentangled, which I think matches the minumum guarantee that releaseing a qubit should not lead to side effects.